pkg/limayaml/defaults.go

```diff
@@ -905,11 +907,97 @@ func NewVMType(driver string) VMType {
+	// existence of "lima.yaml" does not signify existence of the instance,
+	// because the file is created during the initialization of the instance.
+	for _, f := range []string{
+		filenames.HostAgentStdoutLog, filenames.HostAgentStderrLog,
+		filenames.VzIdentifier, filenames.BaseDisk, filenames.DiffDisk, filenames.CIDataISO,
+	} {
+		file := filepath.Join(dir, f)
+		if _, err := os.Lstat(file); !errors.Is(err, os.ErrNotExist) {
+			return true
+		}
+	}
+	return false
+}
+
+func ResolveVMType(y, d, o *LimaYAML, filePath string) VMType {
+	// Check if the VMType is explicitly specified
+	for i, f := range []*LimaYAML{o, y, d} {
+		if f.VMType != nil && *f.VMType != "" && *f.VMType != "default" {
+			logrus.Debugf("ResolveVMType: resolved VMType %q (explicitly specified in []*LimaYAML{o,y,d}[%d])", *f.VMType, i)
+			return NewVMType(*f.VMType)
+		}
+	}
+
+	// If this is an existing instance, guess the VMType from the contents of the instance directory.
+	// Note that the instance directory may be created by a previous version of Lima.
+	if dir, basename := filepath.Split(filePath); dir != "" && basename == filenames.LimaYAML && isExistingInstanceDir(dir) {
+		vzIdentifier := filepath.Join(dir, filenames.VzIdentifier)
+		if _, err := os.Lstat(vzIdentifier); !errors.Is(err, os.ErrNotExist) {
+			logrus.Debugf("ResolveVMType: resolved VMType %q (existing instance, with %q)", VZ, vzIdentifier)
+			return VZ
+		}
+		logrus.Debugf("ResolveVMType: resolved VMType %q (existing instance, without %q)", QEMU, vzIdentifier)
+		return QEMU
+	}
+
+	// Resolve the best type, depending on GOOS
+	switch runtime.GOOS {
+	case "darwin":
+		macOSProductVersion, err := osutil.ProductVersion()
+		if err != nil {
+			logrus.WithError(err).Warn("Failed to get macOS product version")
+			logrus.Debugf("ResolveVMType: resolved VMType %q (default for unknown version of macOS)", QEMU)
+			return QEMU
+		}
+		// Virtualization.framework in macOS prior to 13.5 could not boot Linux kernel v6.2 on Intel
+		// https://github.com/lima-vm/lima/issues/1577
+		if macOSProductVersion.LessThan(*semver.New("13.5.0")) {
+			logrus.Debugf("ResolveVMType: resolved VMType %q (default for macOS prior to 13.5)", QEMU)
+			return QEMU
+		}
+		// Use QEMU if the config depends on QEMU
+		for i, f := range []*LimaYAML{o, y, d} {
+			if f.Arch != nil && !IsNativeArch(*f.Arch) {
+				logrus.Debugf("ResolveVMType: resolved VMType %q (non-native arch=%q is specified in []*LimaYAML{o,y,d}[%d])", QEMU, *f.Arch, i)
+				return QEMU
+			}
+			if f.Firmware.LegacyBIOS != nil && *f.Firmware.LegacyBIOS {
+				logrus.Debugf("ResolveVMType: resolved VMType %q (firmware.legacyBIOS is specified in []*LimaYAML{o,y,d}[%d])", QEMU, i)
+				return QEMU
+			}
+			if f.MountType != nil && *f.MountType == NINEP {
+				logrus.Debugf("ResolveVMType: resolved VMType %q (mountType=%q is specified in []*LimaYAML{o,y,d}[%d])", QEMU, NINEP, i)
+				return QEMU
+			}
+			if f.Audio.Device != nil {
```

Comment:
Not familiar with the audio stuff. Issue/PR is welcome.

Reply:
The audio.device was just copied from video.display

They can return errors in validation, instead of warnings?

Reply:
> They can return errors in validation, instead of warnings?

@afbjorklund Yes, they can. The warning comes from `func (l *LimaVzDriver) Validate() error {` and just write to the log, but don't return an error:

https://github.com/lima-vm/lima/blob/374db8bb91ecd8a3e78bb8d48351fd05e2fa2214/pkg/vz/vz_driver_darwin.go#L134-L139

I think all the warning should return an error instead (but in a separate PR).

I also think `limaDriver.Validate()` should be called as part of `limayaml.Validate()` or through some wrapper that will call both. Calling it as part of `instance.Prepare()` is too late.